Libs/MRML/Core/vtkMRMLBitStreamVolumeNode.cxx

```diff
@@ -0,0 +1,247 @@
+#include "vtkMRMLBitStreamVolumeNode.h"
+
+// MRML includes
+#include "vtkMRMLScene.h"
+#include "vtkMRMLNRRDStorageNode.h"
+
+// VTK includes
+#include <vtkCollection.h>
+#include <vtkMatrix4x4.h>
+#include <vtkNew.h>
+#include <vtkObjectFactory.h>
+#include <vtkXMLUtilities.h>
+
+//----------------------------------------------------------------------------
+vtkMRMLNodeNewMacro(vtkMRMLBitStreamVolumeNode);
+
+//----------------------------------------------------------------------------
+// vtkMRMLBitStreamVolumeNode methods
+
+//-----------------------------------------------------------------------------
+vtkMRMLBitStreamVolumeNode::vtkMRMLBitStreamVolumeNode()
+{
+  this->CompressionDevice = NULL;
+  this->FrameUpdated = false;
+  this->KeyFrameReceived = false;
+  this->KeyFrameDecoded = false;
+  this->KeyFrameUpdated = false;
+  this->FrameMessageValid = false;
+  this->KeyFrameMessage = "";
+  this->FrameMessage = "";
+}
+
+//-----------------------------------------------------------------------------
+vtkMRMLBitStreamVolumeNode::~vtkMRMLBitStreamVolumeNode()
+{
+  if (this->CompressionDevice != NULL)
+    {
+    this->CompressionDevice->Delete();
+    }
+}
+
+//---------------------------------------------------------------------------
+int vtkMRMLBitStreamVolumeNode::ObserveOutsideCompressionDevice(vtkMRMLCompressionDeviceNode* device)
+{
+  if (device)
+    {
+    device->AddObserver(device->GetDeviceContentModifiedEvent(), this, &vtkMRMLBitStreamVolumeNode::ProcessDeviceModifiedEvents);
+    //------
+    if (device->GetContent() != NULL)
+      {
+      std::string bitStream = std::string(device->GetContent()->frameMessage);
+      this->SetFrameMessage(bitStream);
+      if (device->GetContent()->image != NULL)
+        {
+        this->SetAndObserveImageData(device->GetContent()->image);
+        }
+      }
+    this->CompressionDevice = device; // should the interal video device point to the external video device?
+    //-------
+    return 1;
+    }
+  return 0;
+}
+
+void vtkMRMLBitStreamVolumeNode::CopySrcStringToDestString(const std::string& srcString, std::string& destString)
+{
+  destString.resize(srcString.length());
+  char * array = new char[srcString.length()];
+  memcpy(array, srcString.c_str(), srcString.length());
+  destString.assign(array, srcString.length());
+}
+
+void vtkMRMLBitStreamVolumeNode::SetFrameMessage(std::string& buffer)
```

Comment:
It would be nice if we would not be forced to make a deepcopy. For example, if we use vtkUnsignedCharArray (or other VTK object) to store the buffer instead of std::string then we can pass the ownership without actual memory copy.

"Message" should not be included in the name, as we may decode from file. "Set" indicates that previous content is overwritten, while here it is updated/combined/appended. So a name like UpdateFromDataStream or something similar would be better.

Reply:
HI Andras,
i made the frame and the keyframe of type vtkUnsignedCharArray in StreamingVolumeNode.
Also Set/Get frame functions are implemented. So there wouldn't be deep copy among StreamingVolumeNodes in Slicer. 
However, do you think it is better to make the vtkStreamingVolumeCodec use vtkUnsignedCharArray instead of std::string?  Currently i still make copy of the std::string from vtkStreamingVolumeCodec into  StreamingVolumeNode.  As vtkStreamingVolumeCodec is the wrapping class of external codec libraries, the encoded/decode bit stream from external codec libraries will be saved in the std::string structure in vtkStreamingVolumeCodec, deep copy of std::string into StreamingVolumeNode would be safer to avoid the memory segmentation fault.  What do you think? 
Thanks!

Reply:
Unlike vtkObject-based objects, std::string does not give a chance to share or transfer ownership, therefore it gives us more flexibility. You may still DeepCopy a vtkUnsignedCharArray, if needed.

Reply:
vtkUnsignedCharArray is used for all now